### ISSUE TITLE
Multi ROI cell detection fails randomly

### ISSUE THREAD

**Issue by Fafa87** [qupath/qupath#110]
Hello!

We had been running qupath for cell detection on a number of small images crops and it worked well. However later we decided that it would be better to make a mosaic out of input images and open one instance of qupath with a mosaic image and then run cell detection on a number of ROI.

This time it did not work well. At the begining it seemed to work, so we started integrating that. Later it turned out that it randomly messes up giving a result of one ROI in multiple ones. You can see that in the screenshot. 

![image](https://user-images.githubusercontent.com/9865688/31624260-610252b6-b2a2-11e7-86b0-b8c8c08c3e7f.png)

It happens only one in a 2-3 runs. What's more today I've been thinking about how I want to avoid multithreading in some other project (as it is error prone) and realized that it may be the case here. So I changed the number of processors per parallel command (in preferences) and it seems that the problem no longer exists :)

To sum up, multi ROI cell detection fails randomly when multithreading is on.

**Comment by petebankhead** [comment]
Hi, this looks to be the same issue as #74 

It should only affect images that are being read with ImageJ, because of the way in which a cropped region is extracted (when you're unlucky, a second thread might call ```setRoi``` at an inopportune moment).  One potential quick fix in the code could be to make [this method](https://github.com/qupath/qupath/blob/v0.1.2/qupath-extension-ij/src/main/java/qupath/imagej/images/servers/ImageJServer.java#L175) synchronized, although this could be refined a bit further.

To the best of my knowledge, images accessed any other way (including all whole slide images) should be unaffected.

This should be fixed in the next QuPath update.  In the meantime, restricting the number of threads should work too.

**Comment by petebankhead** [comment]
Another workaround in your case could be a script like this:
```groovy
def hierarchy = getCurrentHierarchy()
for (annotation in getAnnotationObjects()) {
    hierarchy.getSelectionModel().setSelectedObject(annotation)
    runPlugin('qupath.imagej.detect.nuclei.PositiveCellDetection', '{"detectionImageBrightfield": "Hematoxylin OD",  "requestedPixelSizeMicrons": 0.5,  "backgroundRadiusMicrons": 8.0,  "medianRadiusMicrons": 0.0,  "sigmaMicrons": 1.5,  "minAreaMicrons": 10.0,  "maxAreaMicrons": 400.0,  "threshold": 0.1,  "maxBackground": 2.0,  "watershedPostProcess": true,  "excludeDAB": false,  "cellExpansionMicrons": 5.0,  "includeNuclei": true,  "smoothBoundaries": true,  "makeMeasurements": true,  "thresholdCompartment": "Nucleus: DAB OD mean",  "thresholdPositive1": 0.2,  "thresholdPositive2": 0.4,  "thresholdPositive3": 0.6,  "singleThreshold": true}');
}
```
Or you could set the number of threads in a script:
```groovy
import qupath.lib.gui.prefs.PathPrefs
PathPrefs.setNumCommandThreads(1)
// Do other things... e.g.
selectAnnotations()
runPlugin('qupath.imagej.detect.nuclei.PositiveCellDetection', '{"detectionImageBrightfield": "Hematoxylin OD",  "requestedPixelSizeMicrons": 0.5,  "backgroundRadiusMicrons": 8.0,  "medianRadiusMicrons": 0.0,  "sigmaMicrons": 1.5,  "minAreaMicrons": 10.0,  "maxAreaMicrons": 400.0,  "threshold": 0.1,  "maxBackground": 2.0,  "watershedPostProcess": true,  "excludeDAB": false,  "cellExpansionMicrons": 5.0,  "includeNuclei": true,  "smoothBoundaries": true,  "makeMeasurements": true,  "thresholdCompartment": "Nucleus: DAB OD mean",  "thresholdPositive1": 0.2,  "thresholdPositive2": 0.4,  "thresholdPositive3": 0.6,  "singleThreshold": true}');
PathPrefs.setNumCommandThreads(Runtime.getRuntime().availableProcessors()) // The default
```

**Comment by petebankhead** [comment]
This should be address in [v0.2.0-m3](https://github.com/qupath/qupath/releases/tag/v0.2.0-m3) (and earlier milestones).
